File: stl_io.py
"""Minimal binary STL writer/reader — no external mesh libraries required.

Triangles are represented as an (N, 3, 3) float64 numpy array:
    triangles[i] = [[x0,y0,z0], [x1,y1,z1], [x2,y2,z2]]
Winding order (right-hand rule) determines the outward normal — callers
are responsible for consistent CCW winding as seen from outside the solid.
"""
import struct
import numpy as np


def compute_normals(triangles):
    v0 = triangles[:, 1, :] - triangles[:, 0, :]
    v1 = triangles[:, 2, :] - triangles[:, 0, :]
    n = np.cross(v0, v1)
    lengths = np.linalg.norm(n, axis=1)
    lengths[lengths == 0] = 1.0
    return n / lengths[:, None]


def write_binary_stl(path, triangles, header=b"nursery pot generator"):
    """triangles: (N,3,3) float64 array."""
    triangles = np.asarray(triangles, dtype=np.float64)
    n_tri = triangles.shape[0]
    normals = compute_normals(triangles)
    with open(path, "wb") as f:
        hdr = header[:80]
        hdr = hdr + b" " * (80 - len(hdr))
        f.write(hdr)
        f.write(struct.pack("<I", n_tri))
        for i in range(n_tri):
            f.write(struct.pack("<3f", *normals[i]))
            for j in range(3):
                f.write(struct.pack("<3f", *triangles[i, j]))
            f.write(struct.pack("<H", 0))


def read_binary_stl(path):
    with open(path, "rb") as f:
        header = f.read(80)
        n_tri = struct.unpack("<I", f.read(4))[0]
        tris = np.zeros((n_tri, 3, 3), dtype=np.float64)
        for i in range(n_tri):
            f.read(12)  # normal, recompute ourselves
            for j in range(3):
                tris[i, j] = struct.unpack("<3f", f.read(12))
            f.read(2)
    return tris


def read_ascii_stl(path):
    verts = []
    tri = []
    with open(path, "r", errors="ignore") as f:
        for line in f:
            line = line.strip()
            if line.startswith("vertex"):
                parts = line.split()
                tri.append([float(parts[1]), float(parts[2]), float(parts[3])])
                if len(tri) == 3:
                    verts.append(tri)
                    tri = []
    return np.array(verts, dtype=np.float64)


def read_stl(path):
    """Auto-detect binary vs ASCII and return (N,3,3) triangle array."""
    with open(path, "rb") as f:
        head = f.read(5)
    if head == b"solid":
        # could still be binary (some exporters wrongly prefix "solid"); check file size
        try:
            tris = read_ascii_stl(path)
            if len(tris) > 0:
                return tris
        except Exception:
            pass
    return read_binary_stl(path)


def mesh_stats(triangles):
    """Rough sanity stats: bounding box, and boundary-edge count (edges used
    by exactly one triangle => open boundary, e.g. an intentional open rim)."""
    triangles = np.asarray(triangles)
    pts = triangles.reshape(-1, 3)
    bbox_min = pts.min(axis=0)
    bbox_max = pts.max(axis=0)

    edge_count = {}
    for tri in triangles:
        for a, b in [(0, 1), (1, 2), (2, 0)]:
            p1 = tuple(np.round(tri[a], 4))
            p2 = tuple(np.round(tri[b], 4))
            key = tuple(sorted([p1, p2]))
            edge_count[key] = edge_count.get(key, 0) + 1
    boundary_edges = sum(1 for v in edge_count.values() if v == 1)
    nonmanifold_edges = sum(1 for v in edge_count.values() if v > 2)
    return {
        "n_triangles": len(triangles),
        "bbox_min": bbox_min,
        "bbox_max": bbox_max,
        "size": bbox_max - bbox_min,
        "boundary_edges": boundary_edges,
        "nonmanifold_edges": nonmanifold_edges,
    }

File: docs/js/stlIO.js
// Minimal binary STL reader/writer — no external mesh libraries.
// Triangles: array of [p0,p1,p2] where each p is [x,y,z].

function computeNormal(p0, p1, p2) {
  const v0 = [p1[0] - p0[0], p1[1] - p0[1], p1[2] - p0[2]];
  const v1 = [p2[0] - p0[0], p2[1] - p0[1], p2[2] - p0[2]];
  const n = [
    v0[1] * v1[2] - v0[2] * v1[1],
    v0[2] * v1[0] - v0[0] * v1[2],
    v0[0] * v1[1] - v0[1] * v1[0],
  ];
  const len = Math.hypot(...n) || 1.0;
  return [n[0] / len, n[1] / len, n[2] / len];
}

export function writeBinarySTL(triangles, headerText = "nursery pot generator") {
  const nTri = triangles.length;
  const buf = new ArrayBuffer(80 + 4 + nTri * 50);
  const dv = new DataView(buf);
  const enc = new TextEncoder();
  const headerBytes = enc.encode(headerText.slice(0, 80));
  new Uint8Array(buf, 0, headerBytes.length).set(headerBytes);
  dv.setUint32(80, nTri, true);

  let offset = 84;
  for (const [p0, p1, p2] of triangles) {
    const n = computeNormal(p0, p1, p2);
    dv.setFloat32(offset, n[0], true); offset += 4;
    dv.setFloat32(offset, n[1], true); offset += 4;
    dv.setFloat32(offset, n[2], true); offset += 4;
    for (const p of [p0, p1, p2]) {
      dv.setFloat32(offset, p[0], true); offset += 4;
      dv.setFloat32(offset, p[1], true); offset += 4;
      dv.setFloat32(offset, p[2], true); offset += 4;
    }
    dv.setUint16(offset, 0, true); offset += 2;
  }
  return buf;
}

export function readBinarySTL(arrayBuffer) {
  const dv = new DataView(arrayBuffer);
  const nTri = dv.getUint32(80, true);
  const triangles = [];
  let offset = 84;
  for (let i = 0; i < nTri; i++) {
    offset += 12; // skip normal
    const pts = [];
    for (let v = 0; v < 3; v++) {
      const x = dv.getFloat32(offset, true); offset += 4;
      const y = dv.getFloat32(offset, true); offset += 4;
      const z = dv.getFloat32(offset, true); offset += 4;
      pts.push([x, y, z]);
    }
    offset += 2; // attribute byte count
    triangles.push(pts);
  }
  return triangles;
}

export function readAsciiSTL(text) {
  const triangles = [];
  let tri = [];
  const lines = text.split("\n");
  for (const raw of lines) {
    const line = raw.trim();
    if (line.startsWith("vertex")) {
      const parts = line.split(/\s+/);
      tri.push([parseFloat(parts[1]), parseFloat(parts[2]), parseFloat(parts[3])]);
      if (tri.length === 3) {
        triangles.push(tri);
        tri = [];
      }
    }
  }
  return triangles;
}

export function readSTL(arrayBuffer) {
  const head = new TextDecoder().decode(arrayBuffer.slice(0, 5));
  if (head === "solid") {
    // Could still be binary (some exporters mislabel); try ASCII first,
    // fall back to binary if it doesn't look right.
    try {
      const text = new TextDecoder().decode(arrayBuffer);
      const tris = readAsciiSTL(text);
      if (tris.length > 0) return tris;
    } catch (e) {
      /* fall through */
    }
  }
  return readBinarySTL(arrayBuffer);
}

export function meshStats(triangles) {
  let bboxMin = [Infinity, Infinity, Infinity];
  let bboxMax = [-Infinity, -Infinity, -Infinity];
  const edgeCount = new Map();
  const keyOf = (p) => `${p[0].toFixed(4)},${p[1].toFixed(4)},${p[2].toFixed(4)}`;

  for (const tri of triangles) {
    for (const p of tri) {
      for (let k = 0; k < 3; k++) {
        if (p[k] < bboxMin[k]) bboxMin[k] = p[k];
        if (p[k] > bboxMax[k]) bboxMax[k] = p[k];
      }
    }
    for (const [a, b] of [[0, 1], [1, 2], [2, 0]]) {
      const ka = keyOf(tri[a]);
      const kb = keyOf(tri[b]);
      const key = ka < kb ? `${ka}|${kb}` : `${kb}|${ka}`;
      edgeCount.set(key, (edgeCount.get(key) || 0) + 1);
    }
  }
  let boundaryEdges = 0, nonmanifoldEdges = 0;
  for (const v of edgeCount.values()) {
    if (v === 1) boundaryEdges++;
    if (v > 2) nonmanifoldEdges++;
  }
  return {
    nTriangles: triangles.length,
    bboxMin,
    bboxMax,
    size: bboxMax.map((v, i) => v - bboxMin[i]),
    boundaryEdges,
    nonmanifoldEdges,
  };
}
